### ISSUE TITLE
How should open-source packages deal with breaking changes in Flutter?

### ISSUE THREAD

**Issue by rrousselGit** [flutter/flutter#32474]
Some recent changes broke an open source package I maintain.
That's fine, and I did the fix.

But how can packages add proper version constraints to prevent that from happening?

I know that we can add the following to `pubspec.yaml`:

```yaml
environment:
  sdk: ">=2.0.0 <3.0.0"
  flutter: <something>
```

But Flutter doesn't use semantic versioning.

So we can't use version constraints similar to:

```yaml
environment:
  sdk: ">=2.0.0 <3.0.0"
  flutter: ">= 1.5.0 <= 1.6.0"
```

because between `1.5.0` and the latest version (currently `1.5.8`), there are multiple breaking changes introduced.

**Comment by tvolkert** [comment]
@rrousselGit our current process is that https://github.com/flutter/flutter/wiki/Changelog calls out which breaking changes landed in which versions, so once a breaking change lands, you should be able to release a patch release in your package that declares the proper upper bound in Flutter.  Then you can release the newer version with the fix and the new Flutter constraints (now with a lower bound).

/cc @kf6gpe for visibility

@sfshaza2 we should add a page about this to flutter.dev.

**Comment by tvolkert** [comment]
The lack of semantic versioning in Flutter does mean that packages have no way a priori to know that their package is safe from breakage.

We didn't do semantic versioning primarily because we've found that the majority of changes are technically breaking changes (as we discover when we issue dev releases), so the versioning would cease to have any meaning (e.g. if we bumped the major version number multiple times a day).

**Comment by rrousselGit** [comment]
I see I missed that the changelog document on which version a breaking change appears.

It may be interesting to add that version number on the announcement made here https://groups.google.com/forum/#!forum/flutter-announce once it lands.

**Comment by rrousselGit** [comment]
@tvolkert Similarly, why isn't `pub.dev` using the stable release of Flutter?

pub.dev currently uses 1.5.8, but stable is at 1.5.4.
So if I publish the package with a correct version constraint to work on `stable` and let users know that it won't work on `master`, pub evaluate it as not working.

Example: https://pub.dev/packages/flutter_hooks#-analysis-tab-

**Comment by dnfield** [comment]
One more plug here for conditional compilation in Dart... https://github.com/dart-lang/sdk/issues/33249

**Comment by localpcguy** [comment]
It is bizarre to me that Flutter does not use semantic versioning. Many other very high profile libaries/frameworks manage to follow semver without "bumping the major version multiple times a day". Either they have a steady release cycle (i.e. new version is cut every 6 weeks with ready features) or hold/batch breaking changes up into a major version bump while allowing non-breaking changes to merge in and keep things moving forward.

Anyways, just my opinion, but this affects way more than open-source packages - it affects all teams using Flutter in production apps. For example (and we are literally going over how we plan to fix this today), we were running `flutter upgrade` in our CI deploy, with the (incorrect) assumption that if major version bumps hadn't happened we could be confident the app would not suffer significant breaking changes. 

I don't think requiring people to constantly scan the Changelog is a particularly effective way of communicating breaking changes, especially since we have things like SemVer which do that much more effectively and in a machine-parsable manner.

**Comment by tvolkert** [comment]
@localpcguy what channel are you on?  The stable channel is only update ~once a quarter.

**Comment by tvolkert** [comment]
Also, for more information on our release process, check out https://github.com/flutter/flutter/wiki/Release-process

**Comment by localpcguy** [comment]
We are on the Stable channel, but that is where we saw breaking changes. That is also why we were surprised. If it is only updated once per quarter (thanks for the Release Process link!), that makes me believe SemVer is a very practical solution - I could see saying only the stable channel follows SemVer. We've fixed at least 3 breaking changes when we tried to go to the latest stable version. I fully admit that likely means we had "bad code" (i.e. if asserts cause breakages, it's probably cause we were not using something correctly), but that still causes build failures and should be consider a breaking change, IMO.

**Comment by tvolkert** [comment]
If you're on the stable channel, then it's safe to assume there will be some breaking changes in every update (about once a quarter if you choose to update).

From our release process, note that the release that makes it to the stable channel was simply a dev channel release that got promoted to beta then promoted to stable - it's version number is decided when it's released to the dev channel.  Based on the cadence of changes (including breaking changes) that land on master, literally _every_ dev channel release would be a major version bump if we followed semver.

@rrousselGit the issue you mention in https://github.com/flutter/flutter/issues/32474#issuecomment-491840580 sounds worth discussing in more detail.  It probably makes make sense as a separate bug so I can loop in the people that manage Pub.  Would you mind filing a separate issue with details as to how the current setup doesn't work?

**Comment by rrousselGit** [comment]
I already opened on the issue on pub repository after contacting other members through gitter.
https://github.com/dart-lang/pub-dartlang-dart/issues/2283

**Comment by stuartmorgan** [comment]
I'm going to close this, as it doesn't seem actionable. The breaking change policy for Flutter is established and documented at this point; while the lack of semver continues to be the case, it doesn't seem to have been a sufficient pain point to necessitate a change to that policy on the last several years.
